src/dai.js

```diff
@@ -0,0 +1,203 @@
+            return df_name.replace('-', '_');
+        }
+        return df_name;
+    }
+
+    _check_parameter() {
+        if (!this.api_url)
+            throw new RegistrationError('api_url is required.');
+
+        if (!this.device_model)
+            throw new RegistrationError('device_model not given.');
+
+        if (this.persistent_binding && !this.device_addr)
+            throw new ArgumentError('In case of `persistent_binding` set to `True`, ' +
+                'the `device_addr` should be set and fixed.');
+
+        if (Object.keys(this.device_features).length === 0)
+            throw new RegistrationError('Neither idf_list nor odf_list is empty.');
+    }
+
+    run() {
+        this._check_parameter();
+
+        this.dan = new Client();
+
+        let idf_list = [];
+        let odf_list = [];
+
+        for (const [df_name, df] of Object.entries(this.device_features)) {
+            if (df.df_type == 'idf')
+                idf_list.push([df_name, df.df_type]);
+            else
+                odf_list.push([df_name, df.df_type]);
+        }
+
+        const msg = {
+            'url': this.api_url,
+            'on_signal': this.on_signal,
+            'on_data': this.on_data,
+            'accept_protos': ['mqtt'],
+            'id': this.device_addr,
+            'idf_list': idf_list,
+            'odf_list': odf_list,
+            'name': this.device_name,
+            'profile': {
+                'model': this.device_model,
+                'u_name': this.username,
+                'extra_setup_webpage': this.extra_setup_webpage,
+                'device_webpage': this.device_webpage,
+            },
+            'register_callback': this.register_callback,
+            'on_register': this.on_register,
+            'on_deregister': this.on_deregister,
+            'on_connect': this.on_connect,
+            'on_disconnect': () => {
+                for (const key in this.flags) {
+                    this.flags[key] = false;
+                }
+                console.debug(`on_disconnect: _flag = ${this.flags}`);
+                if (on_disconnect) {
+                    return on_disconnect;
+                }
+            }
+        };
+
+        this.dan.register(msg);
+
+        // FIXME: window is not defined in node.js
+        window.onbeforeunload = function () {
+            try {
+                if (!this.persistent_binding) {
+                    this.dan.deregister();
+                }
+            } catch (error) {
+                console.error(`dai process cleanup exception: ${error}`);
+            }
+        };
+    }
+
+    parse_df_profile(option, typ) {
+        const df_list = `${typ}_list`;
+        for (let i = 0; i < option[df_list].length; i++) {
+            let df_name;
+            let param_type;
+            let on_data;
+            let push_data;
+            if (!Array.isArray(option[df_list][i])) {
+                df_name = option[df_list][i].name;
```

Comment:
所以做命名規則的轉換是在哪裡？
這邊我看起來沒有做。
```suggestion
                df_name = option[df_list][i].name;
```

Reply:
因為我 df_func_name() 的命名轉換是依照 Python 的方式寫的，所以使用在 push_data() 和 on_data() 中，我會修改成寫在 parse_df_profile() 中，這樣的方式會將 sa 傳進來的  `_[A-Z]?(I|O)[0-9]?` function.name 轉換成 `-[A-Z]?(I|O)[0-9]?`

Reply:
哦 ok